### PR TITLE
Add omit_redundant_param_types option to the javascript generator.

### PR DESCRIPTION
**Pull request by vantreeseba**
- [x] I have read and understood the [Contribution Guidelines](https://github.com/kkoomen/vim-doge/blob/master/CONTRIBUTING.md).
- [x] I have read and understood the [Code of Conduct](https://github.com/kkoomen/vim-doge/blob/master/CODE_OF_CONDUCT.md).

# Why this PR?

This would close issue #68.

It also makes the JSDoc work basically perfectly for haxe as well as typescript.

### CODE REVIEW

**Comment by vantreeseba** [comment]
@kkoomen
Whenever you have some time, I couldn't get the tests to run locally (also I will make another commit with better naming / etc on the test), could you let me know if I am running the tests locally incorrectly?

I see there is a scripts/run-vader-tests.sh and I ran that as
```
./scripts/run-vader-tests.sh vim
```

But it just hung, is there something else I should be doing / need installed to so that I can attempt to clean this PR up and get it to the point where it is acceptable?

**Comment by kkoomen** [comment]
@vantreeseba I don't think it's stuck. It just takes a damn long time to finish. Anyway, you can install the vim plugin Vader locally and then open the test file and run `:Vader` in there. That will be the quickest way to test your own file.

File: README.md [hunk 1]
```diff
@@ -279,12 +279,16 @@ being aliased, see [`g:doge_filetype_aliases`](#gdoge_filetype_aliases).
...
 let g:doge_javascript_settings = {
 \  'destructuring_props': 1,
+\  'omit_redundant_param_types': 0,
 \}
 ```
 
 - `destructuring_props`: Whether or not to generate `@param` tags for the
   destructured properties in a function expression.
 
+- `omit_redundant_param_types`: Whether or not to omit the `{Type}` part of
```
Comment (kkoomen) [inline README.md:289]: - Please use `{type}` where the first letter is lowercase.

File: test/filetypes/javascript/functions.vader [hunk 2]
```diff
@@ -373,6 +373,46 @@ Expect javascript (generated comments with a description, @param and @return tag
    */
   function foo({ a, b: str, c = 3, d: int = 5 }) {}
 
+# ==============================================================================
+# Test the 'omit_redundant_param_types' option
+# ==============================================================================
+Given javascript (functions with type information):
```
Comment (kkoomen) [inline test/filetypes/javascript/functions.vader:379]: Please use the following description:

`functions where the type should be omitted if let g:doge_javascript_settings = {'omit_redundant_param_types': 1} is set`

to make it easier to find this test in the whole series of tests.

**Review by kkoomen — CHANGES_REQUESTED** [review]

**Comment by vantreeseba** [comment]
Both of the requested changes should be implemented, and I did get vader working locally.

Thanks so much for the help.

File: autoload/doge/preprocessors/javascript.vim [hunk 3]
```diff
@@ -55,15 +55,19 @@ function! doge#preprocessors#javascript#tokens(tokens) abort
     if get(g:doge_javascript_settings, 'omit_redundant_param_types') == v:true
       let l:filtered_params = []
       for l:param in a:tokens['parameters']
-        if has_key(l:param, 'type') == v:true && !empty(l:param['type'])
-          let l:param['not_redundant'] = v:false
+        if has_key(l:param, 'type') == v:true
+          if(!empty(l:param['type']))
```
Comment (kkoomen) [inline autoload/doge/preprocessors/javascript.vim:59]: - Please add a space after `if`
- Remove the unnecessary parenthesis around `!empty()`

**Review by kkoomen — CHANGES_REQUESTED** [review]

File: autoload/doge/preprocessors/javascript.vim [hunk 4]
```diff
@@ -52,6 +52,25 @@ endfunction
 " function will adjust the input if needed.
 function! doge#preprocessors#javascript#tokens(tokens) abort
   if has_key(a:tokens, 'parameters') && !empty(a:tokens['parameters'])
+    if get(g:doge_javascript_settings, 'omit_redundant_param_types') == v:true
+      let l:filtered_params = []
```
Comment (kkoomen) [inline autoload/doge/preprocessors/javascript.vim:56]: You can just remove this variable, as you don't use it, right?

**Review by kkoomen — CHANGES_REQUESTED** [review]

**Comment by vantreeseba** [comment]
I realized the behavior of the return types is not correct, I will add another test and get some more code in to correct this.

I.e. Currently, no matter what the return is hidden (when omit_redundant_param_types is true), but it should only be hiding the type template portion is the type is known.

**Comment by kkoomen** [comment]
@vantreeseba Regarding that the `./scripts/run-vader-scripts.sh` is not working locally is because in the CI it does clone the vader inside the current directory instead of being a sibling of the current working directory. I'll try to fix this asap.

**Comment by kkoomen** [comment]
@vantreeseba Please fix tests so I can merge this soon. I want to merge this ASAP.

**Comment by vantreeseba** [comment]
Hmm, I'm not sure why these tests are failing, I ran vader on those files locally and it seems to be working fine.  If you have any insight into why they would be, let me know and I will correct it.

**Comment by kkoomen** [comment]
@vantreeseba The way Vader works when you pass multiple tests is that it will use a single vim instance to run everything. If tests fail, but locally running `:Vader` on the single file that fails, but doesn't fail locally, then the problem is that a combination of previous tests + current test will make the current test fail.

I'll make some time to fix the tests, because it's a bit troublesome to debug this.

**Comment by VVKot** [comment]
Hey @kkoomen / @vantreeseba, wonder if there was any progress here? Anything I can help with?

**Comment by kkoomen** [comment]
@VVKot I have time this weekend, so I can try to fix this.

**Review by kkoomen — APPROVED** [review]

**Comment by kkoomen** [comment]
This feature has been merged and released in [v3.10.0](https://github.com/kkoomen/vim-doge/tree/v3.10.0).

Feel free to submit any new issues if you experience any unwanted behavior in the future. Thanks for your contribution.

**Pull request merged**